File: src/main/java/pl/mjedynak/idea/plugins/pit/actions/PitTestSomeClassesAction.java
package pl.mjedynak.idea.plugins.pit.actions;

import static pl.mjedynak.idea.plugins.pit.cli.model.PitCommandLineArgument.TARGET_CLASSES;
import static pl.mjedynak.idea.plugins.pit.cli.model.PitCommandLineArgument.TARGET_TESTS;

import com.intellij.openapi.fileTypes.StdFileTypes;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.roots.ProjectRootManager;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiManager;
import org.jetbrains.annotations.NotNull;
import pl.mjedynak.idea.plugins.pit.cli.PitCommandLineArgumentsContainer;
import pl.mjedynak.idea.plugins.pit.cli.factory.DefaultArgumentsContainerFactory;
import pl.mjedynak.idea.plugins.pit.cli.factory.DefaultArgumentsContainerPopulator;
import pl.mjedynak.idea.plugins.pit.configuration.PitConfigurationType;
import pl.mjedynak.idea.plugins.pit.configuration.PitRunConfiguration;

/**
 * Runs pitest filtered to a directory of code
 */
public class PitTestSomeClassesAction extends DirectoryOrFilePitAction {

    @Override
    String getTitleForItem(final String item) {
        return "Pitest classes in '" + item + "'";
    }

    @Override
    boolean isEnabled(@NotNull final Project project, @NotNull final Module module, @NotNull final VirtualFile vfile) {
        return (vfile.isDirectory() || (vfile.getFileType() == StdFileTypes.JAVA))
                && module.getModuleContentScope().contains(vfile)
                && !module.getModuleTestsWithDependentsScope().contains(vfile);
    }

    @Override
    PitRunConfiguration makeConfigurationForClassList(
            @NotNull final String classList, @NotNull final Project project, @NotNull final String title) {
        final DefaultArgumentsContainerPopulator defaultArgumentsContainerPopulator =
                new DefaultArgumentsContainerPopulator(
                        ProjectRootManager.getInstance(project), PsiManager.getInstance(project));

        final DefaultArgumentsContainerFactory defaultArgumentsContainerFactory =
                new DefaultArgumentsContainerFactory(defaultArgumentsContainerPopulator) {
                    @Override
                    public PitCommandLineArgumentsContainer createDefaultPitCommandLineArgumentsContainer(
                            final Project proj) {
                        final PitCommandLineArgumentsContainer container =
                                super.createDefaultPitCommandLineArgumentsContainer(proj);
                        container.put(TARGET_CLASSES, classList);
                        container.put(TARGET_TESTS, "*");
                        return container;
                    }
                };

        return new PitRunConfiguration(
                "PIT for classes in " + title,
                project,
                PitConfigurationType.getInstance().getConfigurationFactories()[0],
                defaultArgumentsContainerFactory);
    }
}

File: src/main/kotlin/pl/mjedynak/idea/plugins/pit/actions/PitTestSomeClassesAction.kt
package pl.mjedynak.idea.plugins.pit.actions

import com.intellij.ide.highlighter.JavaFileType
import com.intellij.openapi.module.Module
import com.intellij.openapi.project.Project
import com.intellij.openapi.roots.ProjectRootManager
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.PsiManager
import pl.mjedynak.idea.plugins.pit.cli.PitCommandLineArgumentsContainer
import pl.mjedynak.idea.plugins.pit.cli.factory.DefaultArgumentsContainerFactory
import pl.mjedynak.idea.plugins.pit.cli.factory.DefaultArgumentsContainerPopulator
import pl.mjedynak.idea.plugins.pit.cli.model.PitCommandLineArgument.TARGET_CLASSES
import pl.mjedynak.idea.plugins.pit.cli.model.PitCommandLineArgument.TARGET_TESTS
import pl.mjedynak.idea.plugins.pit.configuration.PitConfigurationType
import pl.mjedynak.idea.plugins.pit.configuration.PitRunConfiguration

class PitTestSomeClassesAction : DirectoryOrFilePitAction() {
    override fun getTitleForItem(item: String): String = "Pitest classes in '$item'"

    override fun isEnabled(
        project: Project,
        module: Module,
        vfile: VirtualFile,
    ): Boolean =
        (vfile.isDirectory || vfile.fileType == JavaFileType.INSTANCE) &&
            module.moduleContentScope.contains(vfile) &&
            !module.moduleTestsWithDependentsScope.contains(vfile)

    override fun makeConfigurationForClassList(
        classList: String,
        project: Project,
        title: String,
    ): PitRunConfiguration {
        val defaultArgumentsContainerPopulator =
            DefaultArgumentsContainerPopulator(
                ProjectRootManager.getInstance(project),
                PsiManager.getInstance(project),
            )
        val defaultArgumentsContainerFactory =
            object : DefaultArgumentsContainerFactory(defaultArgumentsContainerPopulator) {
                override fun createDefaultPitCommandLineArgumentsContainer(`project`: Project): PitCommandLineArgumentsContainer {
                    val container = super.createDefaultPitCommandLineArgumentsContainer(project)
                    container.put(TARGET_CLASSES, classList)
                    container.put(TARGET_TESTS, "*")
                    return container
                }
            }
        return PitRunConfiguration(
            "PIT for classes in $title",
            project,
            PitConfigurationType.getInstance()!!.configurationFactories[0],
            defaultArgumentsContainerFactory,
        )
    }
}
